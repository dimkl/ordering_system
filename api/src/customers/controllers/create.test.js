
const supertest = require("supertest");

const app = require("../../index");
const Customer = require("../models/customer");

describe("POST /customers", () => {
  let request = {};

  beforeAll(() => { request = supertest(app) });
  beforeEach(() => Customer.knex().raw('truncate customers, orders cascade;'));
  afterAll(() => {
    app.close();
    Customer.knex().destroy();
  });

  it("creates and returns a customer", async () => {
    const response = await request.post("/customers")
      .send({
        "first_name": "Dimitris",
        "last_name": "Klouvas",
        "email": "dimitris.klouvas@gmail.com",
        "password": "1234"
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot({
      id: expect.anything()
    });
  });

  it("throws validation error for required properties", async () => {
    const response = await request.post("/customers")
      .send({})
      .set("Accept", "application/json");

    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });

  it("throws validation error for unique email", async () => {
    const email = "customer@example.com";
    await Customer.query().insert({
      first_name: "First",
      last_name: "Customer",
      email: "customer@example.com",
      password: "1234"
    })

    const response = await request.post("/customers")
      .send({
        "first_name": "Dimitris",
        "last_name": "Klouvas",
        "email": email,
        "password": "1234"
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(422);
    expect(response.body).toMatchSnapshot();
  });
});