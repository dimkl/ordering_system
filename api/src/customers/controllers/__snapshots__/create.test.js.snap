// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /customers creates and returns a customer 1`] = `
Object {
  "email": "dimitris.klouvas@gmail.com",
  "first_name": "Dimitris",
  "last_name": "Klouvas",
}
`;

exports[`POST /customers throws validation error for required properties 1`] = `
Array [
  Object {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'first_name'",
    "params": Object {
      "missingProperty": "first_name",
    },
    "schemaPath": "#/required",
  },
  Object {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'last_name'",
    "params": Object {
      "missingProperty": "last_name",
    },
    "schemaPath": "#/required",
  },
  Object {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'email'",
    "params": Object {
      "missingProperty": "email",
    },
    "schemaPath": "#/required",
  },
  Object {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'password'",
    "params": Object {
      "missingProperty": "password",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`POST /customers throws validation error for unique email 1`] = `
Object {
  "message": "email already exists!",
}
`;
